hypothesis-python/tests/conjecture/common.py

```diff
@@ -67,3 +71,153 @@ def accept(f):
             )
 
     return accept
+
+
+def fresh_data():
+    return ConjectureData(BUFFER_SIZE, prefix=b"", random=Random())
+
+
+@st.composite
+def draw_integer_kwargs(
+    draw,
+    *,
+    use_min_value=True,
+    use_max_value=True,
+    use_shrink_towards=True,
+    use_weights=True,
+    use_forced=False,
+):
+    min_value = None
+    max_value = None
+    shrink_towards = 0
+    weights = None
+
+    # this generation is complicated to deal with maintaining any combination of
+    # the following invariants, depending on which parameters are passed:
+    #
+    # (1) min_value <= forced <= max_value
+    # (2) max_value - min_value + 1 == len(weights)
+    # (3) len(weights) <= 1024
+
+    forced = draw(st.integers()) if use_forced else None
+    if use_weights:
+        # handle the weights case entirely independently from the non-weights
+        # case. We'll treat the weight width here as our "master" draw and base
+        # all other draws around that result.
+        assert use_max_value
+        assert use_min_value
+
+        width = draw(st.integers(1, 1024))
+        weights = draw(
+            st.lists(
+                # weights doesn't play well with super small floats.
+                st.floats(
+                    min_value=0.1, max_value=1, allow_nan=False, allow_infinity=False
+                ),
+                min_size=width,
+                max_size=width,
+            )
+        )
+        center = forced if use_forced else draw(st.integers())
+        # pick a random pivot point in the width to split into left and right
+        # segments, for min and max value.
+        pivot = draw(st.integers(0, width - 1))
+        min_value = center - pivot
+        max_value = center + (width - pivot - 1)
+    else:
+        if use_min_value:
+            min_value = draw(st.integers(max_value=forced))
+        if use_max_value:
+            min_vals = []
+            if min_value is not None:
+                min_vals.append(min_value)
+            if forced is not None:
+                min_vals.append(forced)
+            min_val = max(min_vals) if min_vals else None
+            max_value = draw(st.integers(min_value=min_val))
+
+    if use_shrink_towards:
+        shrink_towards = draw(st.integers())
+
+    if forced is not None:
+        assume((forced - shrink_towards).bit_length() < 128)
+
+    return {
+        "min_value": min_value,
+        "max_value": max_value,
+        "shrink_towards": shrink_towards,
+        "weights": weights,
+        "forced": forced,
+    }
+
+
+@st.composite
+def draw_string_kwargs(draw, *, use_min_size=True, use_max_size=True, use_forced=False):
+    intervals = draw(Intervals)
+    alphabet = [chr(c) for c in intervals]
```

Comment:
This could be quite expensive, if we're iterating over much of the unicode range!  I think it's actually cheap at the moment, due to current limitations of `Intervals`,  but we could future-proof this by using `TextStrategy` directly.

Reply:
using `TextStrategy` directly worked great, thanks! Not sure how I overlooked that.